File: cpp_solutions/Array/Find_Peak_Element.cpp
/*
* A peak element is an element that is greater than its neighbors.
*
* Note: binay search
**/

int findPeakElement(vector<int>& nums)
{
	 int start = 0, end = nums.size() - 1, mid=0;
	 
	 while(start<=end){
        mid=(start+end)/2; 
        // important
		if((mid==0||nums[mid]>nums[mid-1]) &&( nums[mid]>nums[mid+1]||mid==nums.size() - 1)){
			return mid;
		}
		else if(mid > 0 && nums[mid]<nums[mid-1]){
			end = mid-1;
		}
		else
		{
			start = mid + 1;
		}
	}
	return mid;
}

File: python_solutions/Array/Find_Peak_Element.py
class Solution:
    def findPeakElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0 or len(nums) == 1:
            return 0
        
        start, end = 0, len(nums)-1
        while start+1<end:
            mid = start + int((end-start)/2)
            if nums[mid] < nums[mid-1]:
                end = mid
            elif nums[mid] < nums[mid+1]:
                start = mid
            else:
                start = mid
                
        if nums[start] < nums[end]:
            return end
        else:
            return start